src/plugins/exercise/commands.js

```diff
@@ -24,6 +24,9 @@ export function insertExercise(change) {
             return !!document.getClosest(last.key, p2 => p1 === p2)
         }) || document
 
+        // Do not insert section in elements like exercise, admonition, etc.
+        if (parent.object !== 'document' && parent.type !== 'section') return
+
```

Comment:
This check should be done in the UI, not here.

Reply:
Why not in both places considering that not anyone will use our front end?

Reply:
Doing it only here will still allow users to press the button, and leave them wandering why nothing happened, so we need these checks in UI anyway. As for this function: commands are supposed to work even if this creates an invalid document, because it allows combining them into more complex changes within `withoutNormalizing`.

Reply:
> we need these checks in UI anyway

Yes, as we decided earlier.

> As for this function: commands are supposed to work even if this creates an invalid document, because it allows combining them into more complex changes within `withoutNormalizing`

This is something which we will probably never use.  Acctually those functions (inserting sections and exercises) are already very complex which is also probably not the best idea.

Reply:
> This is something which we will probably never use.

Maybe, maybe not, we don't know. And should we ever do, keeping things consistent will only benefit us.

> Acctually those functions (inserting sections and exercises) are already very complex which is also probably not the best idea.

They are nowhere near as complex as many Slate's build-in commands.

Reply:
If we will start thinking of every possible case we will never finish version of Adaptarr which is usable by authors.

> They are nowhere near as complex as many Slate's build-in commands.

For me all `insert` functions should only insert elements. So `insertExercise` should just insert:
```
exercise
   problem
      para
```
Other functionalities are just complicating things. Like this with `figure_caption`, because we could handle this in normalization by unwraping invalid elements but `insertExercise` is trying to  handle all possible cases and is failing at this.

Solution is to not complicate things which doesn't have to be complicated.

Reply:
Normalizations exist to transform an invalid document into a valid one, not as helper utilities for transformations. They need to consider what could have caused an invalid document, and what's the simplest thing they can do to fix it. If you start using normalizations as helper utilities for transformations you'll quickly find situations where there are multiple possible “correct” normalizations and you don't know which to do as you lack context, or end up with an editor in which elements move around seemingly at random. Also, doing anything complex in normalizations will likely lead to code duplication, as you would have to re-create the same cases and transformations across many elements, possibly creating even more complex code due to sheer amount of possible child/parent combinations.

And even if you tried, it is impossible to move complexities of `insertExercise` into normalizations. It is written to support selection spanning multiple elements, and even ending at different depths.

Reply:
As for calling `insertExercise` while selection is inside a `figure_caption`. In my opinion it works correctly: it removes `figure_caption` from figure, and wraps it in an exercise and a problem. The reason it doesn't work correctly is that `exercise_problem`'s `child_type_invalid` normalization is run before `figure_caption`'s `parent_type_invalid`.

Reply:
So why then this behaviour is correct for you and doing the same with note or quotation isn't?

Reply:
You're right, we should extract this, let's call it pre-normalization, part from `insertExercise` into a separate command, and then use it also for inserting other complex structures.